core/src/test/java/org/elasticsearch/index/mapper/DocumentMapperParserTests.java

```diff
@@ -0,0 +1,177 @@
+                .put("properties", new MapBuilder<String, Object>()
+                        .put("a", new MapBuilder<String, Object>()
+                                .put("properties", new MapBuilder<String, Object>()
+                                        .put("c", new MapBuilder<String, Object>()
+                                                .put("type", "integer")
+                                                .map())
+                                        .put("b", new MapBuilder<String, Object>()
+                                                .put("type", "float")
+                                                .map())
+                                        .map())
+                                .map())
+                        .map())
+                .map();
+        assertEquals(withoutDots, DocumentMapperParser.removeDotsInFieldNames(withDots));
+    }
+
+    public void testRemoveDotsInFieldNamesConflict1() {
+        Map<String, Object> withDots = new MapBuilder<String, Object>()
+                .put("properties", new MapBuilder<String, Object>()
+                        .put("a.b", new MapBuilder<String, Object>()
+                                .put("type", "float")
+                                .map())
+                        .put("a", new MapBuilder<String, Object>()
+                                .put("properties", new MapBuilder<String, Object>()
+                                        .put("b", new MapBuilder<String, Object>()
+                                                .put("type", "integer")
+                                                .map())
+                                        .map())
+                                .map())
+                        .map())
+                .map();
+
+        IllegalArgumentException e = expectThrows(IllegalArgumentException.class,
+                () -> DocumentMapperParser.removeDotsInFieldNames(withDots));
+        assertEquals("Mapping contains two definitions for field [a.b]", e.getMessage());
+    }
+
+    public void testRemoveDotsInFieldNamesConflict2() {
+        Map<String, Object> withDots = new MapBuilder<String, Object>()
+                .put("properties", new MapBuilder<String, Object>()
+                        .put("a.b", new MapBuilder<String, Object>()
+                                .put("type", "float")
+                                .map())
+                        .put("a", new MapBuilder<String, Object>()
+                                .put("type", "integer")
+                                .map())
+                        .map())
+                .map();
+
+        IllegalArgumentException e = expectThrows(IllegalArgumentException.class,
+                () -> DocumentMapperParser.removeDotsInFieldNames(withDots));
+        assertEquals("Need to create an object mapping called [a] for field [a.b] but this field already exists and is a [integer]",
+                e.getMessage());
+    }
+}
```

Comment:
Is it worth adding a test for adding mappings for `"a.b.c"` and `"a.b.d"` (so two fields with dots in that need to be merged?